### PR TITLE
Make list backfills endpoint use asyncio

### PR DESCRIPTION
**Pull request by dstandish**
This is a sort of hello world / proof of concept for having an route implemented using asyncio.

Depends on https://github.com/apache/airflow/pull/44226

### CODE REVIEW

File: airflow/api_fastapi/common/db/common.py [hunk 1]
```diff
@@ -47,8 +49,29 @@ def your_route(session: Annotated[Session, Depends(get_session)]):
...
+    Dependency for providing a session.
+
+    For non route function please use the :class:`airflow.utils.session.provide_session` decorator.
+
+    Example usage:
+
+    .. code:: python
+
+        @router.get("/your_path")
+        def your_route(session: Annotated[Session, Depends(get_session)]):
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/db/common.py]:
```suggestion
        def your_route(session: Annotated[Session, Depends(get_async_session)]):
```

File: airflow/api_fastapi/common/db/common.py [hunk 2]
```diff
@@ -47,8 +49,29 @@ def your_route(session: Annotated[Session, Depends(get_session)]):
         yield session
 
 
-def apply_filters_to_select(base_select: Select, filters: Sequence[BaseParam | None]) -> Select:
+async def get_async_session() -> Session:
```
Comment (pierrejeambrun) [inline airflow/api_fastapi/common/db/common.py]: I think there is an AsyncSession type or something similar.

**Review by pierrejeambrun — COMMENTED** [review]
Looks nice.

Just a few small suggestions.

Also I realized that we cannot have 'lazy' loading of some attributes at serialization time (when using pydantic), because this will be 'uncontroled async I/O'.

I don't know if you are encountering this on the Backfill model, but basically it means that everything needs to be eager loaded before, so pydantic does not emit uncontrolled request afterwards. (and unfortunatly we cannot `await BackfillResponse.model_validate`)


Also just to mention that there is a `stream` API to iterate over multiple results. That could be interesting to see if this can help us here.

File: airflow/settings.py [hunk 3]
```diff
@@ -498,7 +498,7 @@ def configure_orm(disable_connection_pool=False, pool_class=None):
 
     engine = create_engine(SQL_ALCHEMY_CONN, connect_args=connect_args, **engine_args, future=True)
     async_engine = create_async_engine(SQL_ALCHEMY_CONN_ASYNC, future=True)
-    create_async_session = sessionmaker(
+    session_maker_async = sessionmaker(
```
Comment (pierrejeambrun) [inline airflow/settings.py]: Your name is more explicit, and I understand why legacy names can feel confusing and this one could be better. (We call a `session_maker` `Session` oO => but that's actually a factory for session, so from the client code perspective we are constructing a session and we think we are manipulating a `Session` class for instantiation, that's especially true for the threadlocal registry `scoped_session`, just create a Session but actually things happen under the hood, we are not just instantiating a session and we don't ask people to explicitely use a factory, just the normal `Session`, that's how sqlalchemy docs and many other documents online recommend to do it.)

Anyway, for consistency, I recommend using `AsyncSession`. (same way as `Session` and `NonScopedSession` above).

In the user code this it just becomes `session = AsyncSession()`

Comment (dstandish) [reply airflow/settings.py]: sure

Comment (dstandish) [reply airflow/settings.py]: to clarify, with this

> that's how sqlalchemy docs and many other documents online recommend to do it

you are saying that the SA docs recommend doing it by storing session factory as AsyncSession, like we did with Session?

Comment (dstandish) [reply airflow/settings.py]: oh, yeah, here i guess https://docs.sqlalchemy.org/en/20/orm/session_api.html#session-and-sessionmaker

Comment (pierrejeambrun) [reply airflow/settings.py]: 👍

Comment (dstandish) [reply airflow/settings.py]: done

**Review by pierrejeambrun — APPROVED** [review]
Nice. Feel free to merge after resolving conflicts.

**Comment by pierrejeambrun** [comment]
Note to myself:

Depends on https://github.com/apache/airflow/pull/44267

**Pull request merged**